### DISCUSSION TITLE
using example of FormProvider Performance will not able to trigger validation

### DISCUSSION THREAD

**Discussion by stanleyyuenyiu** [react-hook-form/react-hook-form#2318] [Q&A] [answered]
according to the example of https://react-hook-form.com/advanced-usage/#FormProviderPerformance

if we put in some validation rules, when u click submit, you will not able to trigger the error, as it remember in memo

e.g

`const NestedInput = React.memo(
  (props) => (
    <>
<input name="test" ref={props.register({
    	required:"input.required"
    })} />
 {props.errors.test?.message}
</>
  ),
  (prevProps, nextProps) =>
    prevProps.formState.isDirty === nextProps.formState.isDirty
);

const Container = (props) => {
  const methods = useFormContext();
  return <NestedInput {...methods} {...props}/>;
}

`
and use it with useForm({mode: 'onBlur',  reValidateMode: 'onBlur'});

**Comment by stanleyyuenyiu** [comment]
Because by default formState.isDirty is false, 

and when u click submit without any value, it will false again, which memorised

unless u fill in something, make it to true, then u can go back to false when empty value

i guess the example is totally miss lead peoples, i strongly recommend to update it!

and none of the formstate is useful in this case https://react-hook-form.com/api#formState

every init value is false, but better use null?

**Comment by bluebill1049** [comment]
send us a pr!!!!!!!!!!!

**Comment by stanleyyuenyiu** [comment]
@bluebill1049 sorry are you very angry? so you close the task? as it is an unresolved issue

**Comment by bluebill1049** [comment]
no, I am not angry. you already know the answer from the above statement, send us a PR then. This is not an issue with this library, it's more of a generic React question, I am putting it up there to help others if you think it's wrong or incorrect, send us a PR, I will review and we go from there.  but 11 "!" mark is the same effect as having 1 of them. hahahah

**Comment by stanleyyuenyiu** [comment]
huh? my question is how to resolve it? how do i know the answer? why not open it for everyone to dicuss?

**Comment by bluebill1049** [comment]
well go create a question at the discussion then, this section is for issues.

**Comment by stanleyyuenyiu** [comment]
> mark

yea, im using too strong language, sorry for that

**Comment by bluebill1049** [comment]
can you provide a codesandbox?

**Comment by bluebill1049** [comment, marked as answer]
looks like the validation works: https://codesandbox.io/s/elastic-christian-25m85?file=/src/App.js

**Reply by bluebill1049** [reply]
oh i see the message is not displaying.

**Reply by bluebill1049** [reply]
ok, find the problem: https://codesandbox.io/s/elastic-christian-25m85?file=/src/App.js the example works, but the problem is because formState is wrapped with proxy, so we need to read it to enable the state to be subscribed.

**Reply by stanleyyuenyiu** [reply]
here is the link to reproduce the issue https://codesandbox.io/s/controller-rules-f8i1j?file=/src/App.tsx:0-842

**Reply by bluebill1049** [reply]
that's because we only subscribed to the isDirty change: 
```
(prevProps, nextProps) =>
    prevProps.formState.isDirty === nextProps.formState.isDirty
```
you probably want to do an errors object compare

**Reply by stanleyyuenyiu** [reply]
ur example is not working for validation

**Reply by stanleyyuenyiu** [reply]
so which condition i should use in this case

**Reply by stanleyyuenyiu** [reply]
isValid also not work, as it is false by default, click submit, false again

**Reply by stanleyyuenyiu** [reply]
so do you mean like that -> prevProps.errors.test === nextProps.errors.test

**Reply by bluebill1049** [reply]
yes, that callback.

**Reply by stanleyyuenyiu** [reply]
i will test it now, and try to apply as much as test case i can, i let you know the result later, many thanks

**Reply by bluebill1049** [reply]
look, normally you don't need to do this at all, only for the cause where you need perf improvement.

**Reply by stanleyyuenyiu** [reply]
because, if i have 30 fields , each of those split to a small component and each of those using useFormContext, it will render many time for every component when one of those component trigger validation

**Reply by bluebill1049** [reply]
then don't use FormProvider, keep it simple and don't over complicated things.

**Reply by stanleyyuenyiu** [reply]
yea, but if not using that, pass ref to each small component? not sure will have the same issue or not

**Reply by bluebill1049** [reply]
definitely much lighter... remember RHF doesn't trigger re-render while user is typing.

**Reply by bluebill1049** [reply]
react context is not for free.

**Reply by stanleyyuenyiu** [reply]
yea, you are right, i can give it a try, but i have another issue,  since one of those component using trigger(fieldname), it force parent rerender and then all child component re render again.....

**Reply by stanleyyuenyiu** [reply]
anyway, this discussion can close, many thanks

**Reply by Kaytmazov** [reply]
> because, if i have 30 fields , each of those split to a small component and each of those using useFormContext, it will render many time for every component when one of those component trigger validation

HI. How did you solve this? I have exact the same problem
